docs/Versioning.md

```diff
@@ -55,11 +55,142 @@ By way of example, the `ModelProto.ir_version` MUST be present in every model.
 
 ISSUE: decide and document how we want to handle changes to the format that are forward compatible but will cause the generated code to introduce a breaking change (e.g., changing the data type of a field from int64 to int32, int64 to double or bytes to string).  I would recommend that we never do it after we hit 1.0.0 - prior to that we should be fairly liberal with changes to size (e.g., `int32<>int16`, `float<>double`) and conservative with changes to value space (e.g., `integral<>floatingpoint`, `string<>int`, `scalar<>message`)
 
-ISSUE: define type compatiblity rules either here or under model versioning - probably here
+ISSUE: define type compatibility rules either here or under model versioning - probably here
 
 ## Operator versioning
 
-ISSUE: sort out the design and then document.
+ONNX's operator versioning has semantic meaning: the ONNX operator
+version associated with a model controls the set of ONNX operators it
+may use, and the semantic meaning of these operators.  Operator
+versioning permits ONNX to evolve the semantics of the operators it
+specifies in BC-breaking ways, while preserving backwards-compatibility
+with older exported models (which pin the meaning of operators by
+recording the operator version they target).
+
+The ONNX operator version used by a model is recorded in
+`ModelProto.op_version` and is a simple integer (i.e., it does *not*
+use semantic versioning.)  Each operator version of ONNX is associated
+with some machine-readable metadata which specifies the implementation
+for every operator that is in scope.  When ONNX makes BC-breaking
+changes to the semantics of operators (e.g., by changing the meaning
+of an operator or removing an operator), it increments the ONNX
+operator version.  Additions of new operators do NOT increment the
+ONNX operator version (as they are not BC-breaking changes), nor
+do additions of new optional attributes.
+
+When understanding the semantics of an operator version, it is helpful
+to distinguish an *operator implementation* from an operator name
```

Comment:
Change "operator implementation" to "operator definition" or "operator specification"? For "operator implementation", it means "kernel codes" normally?

Reply:
I would prefer that we simply snap to CS conventions and leverage that intuition.

Operators are declared - their signature and semantics are specified, but their implementation is opaque.
Graphs are defined - their signature and semantics are specified, and their implementation is transparent.

This is consistent with other language specs. In C:
     extern float add(float a, float b);              // declares add as a name that can be referenced
     float add(float a, float b) { return a + b; } // defines the implementation of add 

I prefer declaration, but am fine with specification or contract. Definition implies a specific implementation, which I'm not a fan of.